File: packages/rspack/src/config/zod/node.ts
import { z } from "zod";

export function node() {
	return z.literal(false).or(
		z
			.object({
				__dirname: z
					.boolean()
					.or(z.enum(["warn-mock", "mock", "eval-only"]))
					.optional(),
				__filename: z
					.boolean()
					.or(z.enum(["warn-mock", "mock", "eval-only"]))
					.optional(),
				global: z
					.boolean()
					.or(z.enum(["warn"]))
					.optional()
			})
			.strict()
	);
}

File: crates/rspack_core/src/options/node.rs
#[cfg(allocative)]
use rspack_util::allocative;

#[derive(Debug)]
#[cfg_attr(allocative, derive(allocative::Allocative))]
pub struct NodeOption {
  pub dirname: NodeDirnameOption,
  pub global: NodeGlobalOption,
  pub filename: NodeFilenameOption,
}

#[derive(Debug, Clone, Copy)]
#[cfg_attr(allocative, derive(allocative::Allocative))]
pub enum NodeGlobalOption {
  True,
  False,
  Warn,
}

#[derive(Debug, Clone, Copy)]
#[cfg_attr(allocative, derive(allocative::Allocative))]
pub enum NodeDirnameOption {
  True,
  False,
  WarnMock,
  Mock,
  EvalOnly,
  NodeModule,
}

#[derive(Debug, Clone, Copy)]
#[cfg_attr(allocative, derive(allocative::Allocative))]
pub enum NodeFilenameOption {
  True,
  False,
  WarnMock,
  Mock,
  EvalOnly,
  NodeModule,
}
